### ISSUE TITLE
Low score finetuning from fd5d69c and onwards

### ISSUE THREAD

**Issue by fastroem** [google/automl#316]
I am only able to load a checkpoint for finetuning using the COCO 2017 train and validation sets
up to a certain commit. I am using tensorflow 2.1.0, GPU.

My approach:
1. Generate tfrecords train/val from the current latest commit (Apr 24, 2020) 006668f
2. Setup a control training loading the efficientdet-d0 checkpoint, process few images, and directly evaluate it on the COCO 2017 validation set.

a. Download
https://storage.googleapis.com/cloud-tpu-checkpoints/efficientdet/coco/efficientdet-d0.tar.gz

b. Using any commit up to (and including) commit Mar 31, 2020: 770d450, I get the results which (to me) looks fine: 
--mode=train_and_eval
--training_file_pattern=C:/data/MSCOCO17/tfrecords/train/train*.tfrecord
--validation_file_pattern=C:/data/MSCOCO17/tfrecords/val/val*.tfrecord
--val_json_file=C:/data/MSCOCO17/annotations_trainval2017/annotations/instances_val2017.json
--model_name=efficientdet-d0
--model_dir=C:\Users\XXX\Documents\project\code\model
--ckpt=C:\Users\XXX\Documents\GIT\github\automl\efficientdet\efficientdet-d0
--train_batch_size=8
--eval_samples=5000
--num_examples_per_epoch=10
--num_epochs=1
--hparams="use_bfloat16=false,num_classes=90,moving_average_decay=0"
--use_tpu=False

 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.393
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.544
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.463
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.187
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.743
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.736
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.311
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.442
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.452
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.222
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.748
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.750

c. Now I use any commit later than Mar 31, 2020: 770d450 with the same setup as above and get very low error rates:
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.001
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.003
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.001
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.001
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.003
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.039
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.056
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.056
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.002
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.033
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.131

I have isolated the problem to the following commit, from which, and onward, I always get the poor performance
  Apr 1,  2020: fd5d69c

==> I attempted to set use_native_nms=true to no avail.
==> I have also generated tfrecords with the corresponding commit versions, however, the results remain the same. 

These are the commits I have ran the above setup on trying to isolate the problem.

Apr 24  2020: 006668f -> very low scoring ~0.0xx.
Apr 9,  2020: b8cb565 -> very low scoring ~0.0xx.
Apr 4,  2020: dd7aadc -> very low scoring ~0.0xx.
Apr 1,  2020: ec266c0 -> very low scoring ~0.0xx.
Apr 1,  2020: 108c6f4 -> very low scoring ~0.0xx.
Apr 1,  2020: fd5d69c -> very low scoring ~0.0xx.
Mar 31, 2020: 770d450 -> okay
Mar 29, 2020: c470de8 -> okay
Mar 27, 2020: 3ac0f6b -> okay

==> Does anyone have any insights to what is going on here? 
==> Did I configure something wrong in my setup?
==> Are the provided ckpt not valid from fd5d69c and onward? If so, could you provide updated ones?

Thanks in advance, any input is appreciated.

**Comment by fsx950223** [comment]
Change max_boxes_to_draw to a very big value could fix the bug.
It seems that you are using tf postprocessing for mAP test?

**Comment by mingxingtan** [comment]
I just checkout the latest code and checkpoints, and I am able to reproduce the coco eval accuracy. Here are the command lines (same as the [README](https://github.com/google/automl/blob/master/efficientdet/README.md)):

```
// Download coco data.
!wget http://images.cocodataset.org/zips/val2017.zip
!wget http://images.cocodataset.org/annotations/annotations_trainval2017.zip
!unzip val2017.zip
!unzip annotations_trainval2017.zip

// convert coco data to tfrecord.
!mkdir tfrecord
!PYTHONPATH=".:$PYTHONPATH"  python dataset/create_coco_tfrecord.py \
    --image_dir=val2017 \
    --caption_annotations_file=annotations/captions_val2017.json \
    --output_file_prefix=tfrecord/val \
    --num_shards=32

// Run eval.
!python main.py --mode=eval  \
    --model_name=${MODEL}  --model_dir=${CKPT_PATH}  \
    --validation_file_pattern=tfrecord/val*  \
    --val_json_file=annotations/instances_val2017.json  \
    --hparams="use_bfloat16=false" --use_tpu=False
```

**Comment by fastroem** [comment]
Thanks for checking @mingxingtan 
However, my problem is a bit different than your approach. Instead of using 
  --mode=eval, 
try using 
  --mode=train_and_eval 
and start training from the checkpoint, then directly evaluate, and check the error values.

**Comment by elesun2018** [comment]
when i run eval code on coco with pretrained models, meet a problem as follows:
automl0401/efficientdet# python main.py --mode=eval --model_name=${MODEL} --model_dir=${CKPT_PATH} --validation_file_pattern=tfrecord/val* --val_json_file=annotations/instances_val2017.json --hparams="use_bfloat16=false" --use_tpu=False

CKPT_PATH is the pretrained d0 model.

DONE (t=34.86s).
Average Precision (AP) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.002
Average Precision (AP) @[ IoU=0.50 | area= all | maxDets=100 ] = 0.005
Average Precision (AP) @[ IoU=0.75 | area= all | maxDets=100 ] = 0.002
Average Precision (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.003
Average Precision (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.001
Average Precision (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 1 ] = 0.006
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 10 ] = 0.011
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.012
Average Recall (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.018
Average Recall (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.006
Average Recall (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.004
INFO:tensorflow:Finished evaluation at 2020-04-28-17:12:04
INFO:tensorflow:Finished evaluation at 2020-04-28-17:12:04
I0428 17:12:04.599111 140494972888896 evaluation.py:275] Finished evaluation at 2020-04-28-17:12:04
INFO:tensorflow:Saving dict for global step 0: AP = 0.0022378154, AP50 = 0.0052162167, AP75 = 0.0015809378, APl = 0.00015473206, APm = 0.0014778649, APs = 0.0029822278, ARl = 0.0036031725, ARm = 0.006258475, ARmax1 = 0.005900326, ARmax10 = 0.011014862, ARmax100 = 0.011780873, ARs = 0.018089531, box_loss = 0.0, cls_loss = 0.02997985, global_step = 0, loss = 0.11981841
I0428 17:12:04.600037 140494972888896 estimator.py:2049] Saving dict for global step 0: AP = 0.0022378154, AP50 = 0.0052162167, AP75 = 0.0015809378, APl = 0.00015473206, APm = 0.0014778649, APs = 0.0029822278, ARl = 0.0036031725, ARm = 0.006258475, ARmax1 = 0.005900326, ARmax10 = 0.011014862, ARmax100 = 0.011780873, ARs = 0.018089531, box_loss = 0.0, cls_loss = 0.02997985, global_step = 0, loss = 0.11981841
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 0: models/pretrained_models/efficientdet-d1/model

such low AP values ? why ?
thanks.

**Comment by zishanahmed08** [comment]
same issue here
